### ISSUE TITLE
How to embed native control or get native handle of existing control

### ISSUE THREAD

**Issue by visioforge** [AvaloniaUI/Avalonia#2131]
I'm working on video player (GStreamer based) and want to render video inside Avalonia app.

I'm using Panel in WinForms app as video window. 

I need some area inside Avalonia window (Panel for example) that can provide native window handle to draw on it. It's possible for Win32 control or Gtk controls for example. 

I can get handle for entire Window but that's not what I need.
 
`var screenHandle = ((Avalonia.Win32.WindowImpl) ((TopLevel) pnScreen.GetVisualRoot())?.PlatformImpl)?.Handle?.Handle ?? IntPtr.Zero;`

As alternative maybe some way to embed native platform controls that have handle into Avalonia window.

**Comment by danwalmsley** [comment]
Avalonia doesnt have native controls so no control has a handle, you can only get a handle to the Window.

**Comment by danwalmsley** [comment]
btw have you seen

https://github.com/jkoritzinsky/AvaloniaAV

not sure what state its in, but might be useful to you.

**Comment by visioforge** [comment]
AvaloniaAV is useless and Windows-only, because based on Media Foundation. My player based on GStreamer and Avalonia is one of UI frameworks used for demo apps. 

Right now I'm using Image control for custom frame rendering and it's not so bad for performance.

But, some way to add native control would be great.

Maybe need to made some custom control with Direct2D rendering. Not sure.

**Comment by Gillibald** [comment]
Why do you want native controls instead of Avalonia controls?

**Comment by Wyvernius** [comment]
In my case it would be usefull to have some "Media Element" with a native handle so can pass this to VLC to draw a "preview".
Then when resizing the control, it would also size the video, etc...

**Comment by danwalmsley** [comment]
@Wyvernius something to do this is coming shortly.

**Comment by Wyvernius** [comment]
@danwalmsley  awsome!
In the meantime i created my own window implementation.
To get a handle inside an UserControl:
```
 public class VlcPreviewWindow : UserControl
    {
        Window window;
        Window Parent;
        public VlcPreviewWindow()
        {
            this.InitializeComponent();
            this.LayoutUpdated += VlcPreviewWindow_LayoutUpdated;
            this.AttachedToVisualTree += VlcPreviewWindow_AttachedToVisualTree;
            window = new Window();
            window.Background = Avalonia.Media.Brushes.Black;
            window.WindowStartupLocation = WindowStartupLocation.Manual;
            window.WindowState = WindowState.Normal;
            window.HasSystemDecorations = false;
            window.ShowInTaskbar = false;
        }

        private void VlcPreviewWindow_AttachedToVisualTree(object sender, VisualTreeAttachmentEventArgs e)
        {
            Parent = (Window)e.Root;
            Parent.PositionChanged += Parent_PositionChanged;
        }

        private void Parent_PositionChanged(object sender, PixelPointEventArgs e)
        {
            window.Position = this.PointToScreen(this.Bounds.Position);
        }

        private void VlcPreviewWindow_LayoutUpdated(object sender, System.EventArgs e)
        {
            window.Position = this.PointToScreen(this.Bounds.Position);
            window.Width = this.Bounds.Width;
            window.Height = this.Bounds.Height;
            window.Topmost = true;
            //NativeMethods.SetWindowPos(HWND, NativeMethods.HWND_TOPMOST, window.Position.X, window.Position.Y, (int)window.Width, (int)window.Height, 0);
        }

        private void InitializeComponent()
        {
            AvaloniaXamlLoader.Load(this);
        }

        public WindowState WindowState
        {
            get { return window.WindowState; }
            set { window.WindowState = value; }
        }

        public void Show()
        {
            window.Show();
        }

        public IntPtr Handle
        {
            get { return window.PlatformImpl.Handle.Handle; }
        }

        public void Close()
        {
            window.Close();
        }
    }
```

**Comment by seandillon92** [comment]
@danwalmsley any updates on this? I am looking into hosting a Unity executable in avalonia controls (unfortunately windows are not an option)

**Comment by kekekeks** [comment]
0.10 will have an equivalent for HwndHost, see https://github.com/AvaloniaUI/Avalonia/blob/master/samples/interop/NativeEmbedSample/EmbedSample.cs

**Comment by raidenXR** [comment]
is it possible to embed _SDL2_ windows on _AvaloniaUI_? I did try to do that on WPF with **HwndHost**, but the _BuildWindowCore()_ throws an _InvalidOperationException_ when returning the _HandleRef_.

**Comment by maxkatz6** [comment]
NativeControlHost was implemented in 0.10
